src/Grpc.AspNetCore.Server/Internal/GrpcServiceBinder.cs

```diff
@@ -31,40 +35,106 @@ namespace Grpc.AspNetCore.Server.Internal
+            _serviceInvokerFactory = serviceInvokerFactory;
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, ClientStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateClientStreaming(method, _serviceInvokerFactory.CreateClientStreamingInvoker(method));
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, DuplexStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateDuplexStreaming(method, _serviceInvokerFactory.CreateDuplexStreamingInvoker(method));
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, ServerStreamingServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateServerStreaming(method, _serviceInvokerFactory.CreateServerStreamingInvoker(method));
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
         }
 
         public override void AddMethod<TRequest, TResponse>(Method<TRequest, TResponse> method, UnaryServerMethod<TRequest, TResponse> handler)
         {
             var callHandler = _serverCallHandlerFactory.CreateUnary(method, _serviceInvokerFactory.CreateUnaryInvoker(method));
-            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, callHandler.HandleCallAsync));
+            AddMethodCore(method, callHandler.HandleCallAsync);
+        }
+
+        private void AddMethodCore(IMethod method, RequestDelegate requestDelegate)
+        {
+            ServiceMethods.Add(method);
+            EndpointConventionBuilders.Add(_builder.MapPost(method.FullName, requestDelegate, new object[] { method }));
+        }
+
+        internal void CreateUnimplementedEndpoints()
+        {
+            // Return UNIMPLEMENTED status for requests with a URL that has two segments and a grpc/application content-type
+
+            // Return UNIMPLEMENTED status for missing service:
+            // - /{service}/{method}
+            if (_serviceMethodsRegistry.Methods.Count == 0)
+            {
+                EndpointConventionBuilders.Add(CreateUnimplementedEndpoint("{unimplementedService}/{unimplementedMethod}", _serverCallHandlerFactory.CreateUnimplementedService()));
+            }
+
+            // Return UNIMPLEMENTED status for missing method:
```

Comment:
comment accuracy: looks like the content type constraint is being applied for unimplemented methods too but from the comment it doesn't look it.